src/util/heap/freelistpageresource.rs

```diff
@@ -191,12 +199,30 @@ impl<VM: VMBinding> FreeListPageResource<VM> {
             meta_data_pages_per_region,
             sync: Mutex::new(FreeListPageResourceSync {
                 pages_currently_on_freelist: 0,
-                highwater_mark: 0,
+                highwater_mark: UNINITIALIZED_WATER_MARK,
             }),
             _p: PhantomData,
+            protect_memory_on_release: false,
         }
     }
 
+    /// Protect the memory
+    fn mprotect(&self, start: Address, pages: usize) {
+        // No `unwrap()` here since this may cause ENOMEM error and fail to protect the memory.
```

Comment:
If mprotect fails, the page is not actually protected. So the GC cannot guarantee that dead objects are always protected and not accessible. 

I tried to get some number on this. I am running `antlr` with PageProtect GC on OpenJDK.
The first failed mprotect happens at ~30k protect. Considering the default limit for mapping is 65535, that sounds reasonable. The ratio of successful/total mprotect attempts is ~15%. That means only 15% of the protection attempts are successfull. That does not necessarily mean 85% of the objects are not protected. But it could be the case that a large proportion of the objects are not protected.

And moreover, what if the protect is successful but the unprotect fails, then we have protected pages that will cannot be accessed. I modified the code to ignore protect fail, but panic for unprotect fails, and the panic actually happened.

Reply:
The 64K mapping limit seems like a real issue here.
What is we have more objects than that? I think this is not uncommon, right?

Reply:
I made these changes:
* have a warning when mmtk is booted with PageProtect to warn people that mapping limit may cause protection fail
* Panic if protect/unprotect fail
* Set higher mapping limit in our test scripts (runs on Github hosted runners with priviledge to do so).